### PR TITLE
feat: Support (De)Serialization for different representations of Nullable Unions

### PR DESCRIPTION
**Pull request by PookieBuns**
Fixes `The Rust Representation Conundrum` as mentioned in #449, which allows (de)serializing, all `avro-rs`, `rusty`, and `avro_json_encoding_compatible` rust representations to avro. Tested using the provided test suite in the mentioned issue.

**Note: This does not address #365 since this only applies to schema aware serialization.**

I considered using enum variant name as lookup method for picking Union branch when serializing, shown [here](https://github.com/PookieBuns/avro-rs/commit/74172ecf3916ec1c82a5dcbca2c8caa0f8c943a7), but given the current state in which `Value` doesn't hold the name for named types, I realized it's not possible to do the same for deserialization until further work to support names in `Value` has been done.

Therefore, this implementation makes the assumption that if you are using `Option<T>` to represent a nullable union, that means that the `null` is the first variant in the union and serialization / deserialization just bump / decrement the index lookup respectively.

I believe this doesn't actually produce any breaking changes, because the status quo is that tagged Option "Rusty" representations straight up don't work.

Please let me know if you have any questions!

### CODE REVIEW

File: avro/src/serde/de.rs [hunk 1]
```diff
@@ -638,6 +664,21 @@ impl<'de> de::Deserializer<'de> for Deserializer<'de> {
...
+                if matches!(x.deref(), Value::Union(_, _)) {
+                    Err(de::Error::custom(format!(
+                        "Directly nested union types are not supported. Got Value::Union({i}, {x:?})"
+                    )))
+                } else {
+                    Self::new(x.deref())
+                        .deserialize_byte_buf(visitor)
+                        .map_err(|e| {
+                            de::Error::custom(format!(
+                                "Attempted to deserialize Value::Union({i}, {x:?}) as bytes: {e:?}"
```
Comment (martin-g) [inline avro/src/serde/de.rs]:
```suggestion
                                "Attempted to deserialize Value::Union({i}, {x:?}) as byte_buf: {e:?}"
```

File: avro/src/serde/de.rs [hunk 2]
```diff
@@ -799,9 +850,19 @@ impl<'de> de::Deserializer<'de> for Deserializer<'de> {
...
-                if (*idx as usize) < variants.len() {
+                // if we came here from a some, we need to check if we are deserializing a
+                // non-newtype enum
+                if self.deserializing_some
+                    && let Value::Enum(_index, field) = inner.deref()
+                    && variants.contains(&&**field)
+                {
+                    return visitor.visit_enum(EnumUnitDeserializer::new(field));
+                }
+                // Assume `null` is the first branch if deserializing some so decrement the variant index
```
Comment (martin-g) [inline avro/src/serde/de.rs:861]: This assumption is not guaranteed by the spec :-/
We should add some index/length checks to avoid panics/overflows.

File: avro/src/serde/ser_schema/mod.rs [hunk 3]
```diff
@@ -1460,19 +1477,37 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
...
+                                    )));
+                                }
+                                encode_int(i as i32, &mut self.writer)?;
+                                return encode_int(variant_index as i32, &mut self.writer);
+                            }
+                            _ => { /* skip */ }
+                        }
+                    }
+                }
+                let branch_index = variant_index as usize + usize::from(self.serializing_some);
```
Comment (martin-g) [inline avro/src/serde/ser_schema/mod.rs:1517]: Again, this assumes that the `null` is the first variant in the union.

File: avro/tests/nullable_union.rs [hunk 4]
```diff
@@ -0,0 +1,982 @@
...
+
+mod nullable_primitive_int {
+    use super::*;
+
+    const NULLABLE_INT_SCHEMA: &str = r#"
+    {
+        "name": "MyUnion",
+        "type": [
+            "null",
+            "int"
```
Comment (martin-g) [inline avro/tests/nullable_union.rs]: Please add a test where the `null` is not the first variant too.

File: avro/src/serde/ser_schema.rs [hunk 5]
```diff
@@ -1524,17 +1559,21 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
...
-                    .ok_or_else(|| {
-                        create_error(format!(
-                            "No variant schema at position {variant_index} for {union_schema:?}"
-                        ))
-                    })?;
+                let branch_index = variant_index as usize + usize::from(self.serializing_some);
+                if branch_index >= union_schema.schemas.len() {
+                    return Err(create_error(format!(
+                        "Variant index out of bounds: {}. The union schema has '{}' schemas",
+                        variant_index,
```
Comment (martin-g) [inline avro/src/serde/ser_schema.rs]:
```suggestion
                        branch_index,
```

File: avro/src/serde/ser_schema.rs [hunk 6]
```diff
@@ -1703,17 +1742,21 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
...
-                    .ok_or_else(|| {
-                        create_error(format!(
-                            "Cannot find a variant at position {variant_index} in {union_schema:?}"
-                        ))
-                    })?;
+                let branch_index = variant_index as usize + usize::from(self.serializing_some);
+                if branch_index >= union_schema.schemas.len() {
+                    return Err(create_error(format!(
+                        "Variant index out of bounds: {}. The union schema has '{}' schemas",
+                        variant_index,
```
Comment (martin-g) [inline avro/src/serde/ser_schema.rs]:
```suggestion
                        branch_index,
```

File: avro/src/serde/ser_schema.rs [hunk 7]
```diff
@@ -1835,17 +1878,17 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
...
-                    .ok_or_else(|| {
-                        create_error(format!(
-                            "Cannot find variant at position {variant_index} in {union_schema:?}"
-                        ))
-                    })?;
+                let branch_index = variant_index as usize + usize::from(self.serializing_some);
+                if branch_index >= union_schema.schemas.len() {
+                    return Err(create_error(format!(
+                        "Variant index out of bounds: {}. The union schema has '{}' schemas",
+                        variant_index,
```
Comment (martin-g) [inline avro/src/serde/ser_schema.rs]:
```suggestion
                        branch_index,
```

File: avro/src/serde/ser_schema/mod.rs [hunk 8]
```diff
@@ -1460,19 +1477,37 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
...
+                                if variant_index as usize >= enum_schema.symbols.len() {
+                                    return Err(create_error(format!(
+                                        "Variant index out of bounds: {}. The Enum schema has '{}' symbols",
+                                        variant_index,
+                                        enum_schema.symbols.len()
+                                    )));
+                                }
+                                encode_int(i as i32, &mut self.writer)?;
+                                return encode_int(variant_index as i32, &mut self.writer);
+                            }
```
Comment (martin-g) [inline avro/src/serde/ser_schema/mod.rs:1496]: This should probably support Schema::Ref too, to refer to Schema::Enum

File: avro/src/serde/de.rs [hunk 9]
```diff
@@ -75,7 +76,17 @@ struct UnionDeserializer<'de> {
...
 impl<'de> Deserializer<'de> {
     pub fn new(input: &'de Value) -> Self {
-        Deserializer { input }
+        Deserializer {
+            input,
+            deserializing_some: false,
+        }
+    }
+
+    pub fn new_deserializing_some(input: &'de Value) -> Self {
```
Comment (martin-g) [inline avro/src/serde/de.rs]:
```suggestion
    fn new_deserializing_some(input: &'de Value) -> Self {
```

File: avro/src/serde/ser_schema.rs [hunk 10]
```diff
@@ -629,6 +630,22 @@ impl<'s, W: Write> SchemaAwareWriteSerializer<'s, W> {
             root_schema: schema,
             names,
             enclosing_namespace,
+            serializing_some: false,
+        }
+    }
+
+    pub fn new_serializing_some(
```
Comment (martin-g) [inline avro/src/serde/ser_schema.rs]:
```suggestion
    fn new_serializing_some(
```

File: avro/tests/nullable_union.rs [hunk 11]
```diff
@@ -0,0 +1,982 @@
...
+    fn serialize_rusty_untagged_my_record_b_27() -> AvroResult<()> {
+        test_serialize(TestCase {
+            input: Some(MyUnionUntagged::MyRecordB(MyRecordB { a: 27 })),
+            schema: &schema(),
+            expected_avro: &record_b_variant_expected_avro(27),
+        })
+    }
+
+    #[test]
+    #[ignore]
```
Comment (martin-g) [inline avro/tests/nullable_union.rs]: Please add a comment why it is ignored

Comment (PookieBuns) [reply avro/tests/nullable_union.rs]: Updated to instead expect a failure / comment on why it will fail. I'm also open to just removing it if you think its out of scope, since that mod is mostly to demonstrate how untagged enums have edge cases that fail and check that tagged enums do not

File: avro/src/serde/de.rs (hunk 2 above)
Comment (PookieBuns) [reply avro/src/serde/de.rs:861]: Updated to use checked_sub to ensure underflow doesn't happen.

Regarding the topic about not being enforced by spec, even though it isn't strictly enforced, it's highly recommended https://avro.apache.org/docs/1.11.1/specification/#unions

> Note that when a default value is specified for a record field whose type is a union, the type of the default value must match the first element of the union. Thus, for unions containing “null”, the “null” is usually listed first, since the default value of such unions is typically null.

For serialization we technically could get around this by resolving union branches using variant name, but for deserialization cannot yet because it's not schema aware and/or Value doesn't hold name information, which is why I chose to use index bumping on serialization too for consistent behavior. This seems like the closest we can get without a significant refactor or feature for deserialization.

For now, the behavior will intentionally throw an error if you try to (de)serialize an Option<T> with null not being the first branch. This doesn't break existing behavior because the status quo is non-functional so although not a perfect solution imo it's a step better than what we have now

File: avro/src/serde/ser_schema/mod.rs (hunk 3 above)
Comment (PookieBuns) [reply avro/src/serde/ser_schema/mod.rs:1517]: Updated to throw error if null isn't first branch, reasoning stated in https://github.com/apache/avro-rs/pull/458#discussion_r2789323892

File: avro/tests/nullable_union.rs (hunk 4 above)
Comment (PookieBuns) [reply avro/tests/nullable_union.rs]: Added tests in https://github.com/apache/avro-rs/pull/458/changes/6f691dbbcf2211eb5c112cda90759f7f76c0d44c to ensure null as non-first variant when using Option<T> fails with error message communicating intended behavior

File: avro/src/serde/ser_schema.rs (hunk 5 above)
Comment (PookieBuns) [reply avro/src/serde/ser_schema.rs]: Updated!

File: avro/src/serde/ser_schema.rs (hunk 7 above)
Comment (PookieBuns) [reply avro/src/serde/ser_schema.rs]: Updated!

File: avro/src/serde/ser_schema/mod.rs (hunk 8 above)
Comment (PookieBuns) [reply avro/src/serde/ser_schema/mod.rs:1496]: updated in https://github.com/apache/avro-rs/pull/458/changes/6c58e62db077ad87efb8549e4200a20f5298b465 with test to verify behavior

File: avro/src/serde/de.rs [hunk 12]
```diff
@@ -799,14 +850,26 @@ impl<'de> de::Deserializer<'de> for Deserializer<'de> {
...
+                if self.deserializing_some
+                    && let Value::Enum(_index, field) = inner.deref()
+                    && variants.contains(&&**field)
+                {
+                    return visitor.visit_enum(EnumUnitDeserializer::new(field));
+                }
+                // Assume `null` is the first branch if deserializing some so decrement the variant index
+                let variant_idx = (*idx as usize)
+                    .checked_sub(usize::from(self.deserializing_some))
+                    .ok_or_else(|| Details::GetNull(inner.deref().clone()))?;
```
Comment (PookieBuns) [inline avro/src/serde/de.rs:864]: Do we want to add an Error variant here to communicate more clearly that deserializing some when null isn't first isn't supported? Right now it just tells the user Expects null for 0th when deserializing some

**Comment by PookieBuns** [comment]
Updated to use new methods introduced by previous commits
